### PR TITLE
Visualize centerline and laplace

### LINKED ISSUE #27: visualize centerline and laplace solution isosurfaces
The VPAWVisualize module should have functionality to visualize the p3 files generated by the [pediatric_airway_atlas](https://github.com/uncbiag/pediatric_airway_atlas) pipelines.

These would be files with the following name pattern:
```txt
./transformed_landmarks/*_LANDMARKS.p3
./areas/*_AREAS.p3
./processed_areas/*_AREAS.p3
./interpolated_areas/*_AREAS.p3
./processed_centerline/*_CENTERLINE.p3
./centerline/*_CENTERLINE.p3
```

UPDATE: See [comment below](https://github.com/KitwareMedical/vpaw/issues/27#issuecomment-1456495085) for the suggested visualization to create here.

### LINKED ISSUE #28: VPAWVisualize does not detect missing patient ID
When inputting missing patient ID, it proceeds to load some plots anyway, with no error message

### PR DESCRIPTION
**Pull request by ebrahimebrahim**
This is primarily work to close #27 

- Centerline points are used to construct a markups curve
- Laplace solution isosurfaces are constructed via marching cubes

Some remaining work before this PR is ready:
- The centerline and the laplace solution are not aligned with everything else (and not even with each other). I still need to determine the coordinate system these images get saved in back in [pediatric_airway_atlas](https://github.com/uncbiag/pediatric_airway_atlas) and then do the right transform to achieve alignment.
- Mask the laplace soln image by the airway segmentation before we create the isosurfaces (need them aligned first!)
- Add a progress indicator for isosurface generation, as it is a bit time consuming.

---

Other minor changes along the way:
- disable loading in plots that are not patient specific
- update slicer
- remove "load all patients" feature as it's not useful and complicates the introduction of new features
- fix #28 and raise an error when a patient is not found

---

To show the extent/type of misalignment:

image

### CODE REVIEW

**Comment by ebrahimebrahim** [comment]
The above TODOs are completed:

- centerline and laplace solution are now in the right coordinate system
- laplace solution is masked by airway segmentation
- isosurface generation has a progress indicator

Plus some more improvements:
- removed need for MergeModels CLI module, which was slowing things down
- massively sped up isosurface computation by setting up vtk filters to do it ourselves. this allowed us to take advantage of the fact that `vtkFlyingEdges3D` can compute multiple contour values at once.
- added spinbox for controlling number of isosurfaces
- added slider for segmentation opacity
- improved default configuration of 3D view: black background, hide coordinate box, show RAS axes, set camera to a reasonable viewpoint.

image

**Comment by ebrahimebrahim** [comment]
@Leengit This is ready to be reviewed. A quick look or a deep review is okay, whatever you prefer.

File: Modules/Scripted/VPAWVisualize/vpawvisualizelib/isosurfaces.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+            should result in one isosurface
+        decimate_target_reduction: by how much to decimate after doing vtkFlyingEdges3D
+        progress_callback: Optionally, a function that takes a progress_percentage float value.
+            If this is provided then progress_callback(progress_percentage) will be called
+            while the computation is being done.
+    Return: a vtkMRMLModelNode
+    """
+
+    if progress_callback is None:
+        progress_callback = lambda progress_percentage : None
```
Comment (Leengit) [inline Modules/Scripted/VPAWVisualize/vpawvisualizelib/isosurfaces.py:20]: This is probably okay as is, but `flake8` doesn't like that a lambda is being assigned a name.  It prefers that we use `def` to assign names to functions.  Maybe this isn't the situation that `flake8` means to catch, but I thought to mention it anyway.

**Review by Leengit — APPROVED** [review]
Looks good.  Small comments inline.  Also, you might run
```bash
black -C Modules/Scripted/VPAWModel/VPAWModel.py \
    Modules/Scripted/VPAWVisualize/VPAWVisualize.py \
    Modules/Scripted/VPAWVisualize/vpawvisualizelib/isosurfaces.py
```
to use the standard formatting defined by `black`.  That won't re-wrap comments to a line length of 88 or less, so you'd have to do that manually.  You can find those lines with something like:
```bash
cd git/KitwareMedical/vpaw
flake8 . --count --select=E9,F63,F7,F82 --show-source --statistics; \
    flake8 . --count --exit-zero  --ignore=E203,W503 --max-complexity=10 --max-line-length=88 --statistics
```

File: Modules/Scripted/VPAWVisualize/VPAWVisualize.py [hunk 2]
```diff
@@ -703,8 +891,51 @@ def arrangeView(self):
...
+        Raises exception if none exists.
+
+        Args:
+            num_isosurface_values: number of isosurface values
+            progress_callback: Optionally, a function that takes a progress_percentage float value.
+                If this is provided then progress_callback(progress_percentage) will be called by compute_isosurfaces
+                while the computation is being done.
+        """
+        if self.laplace_sol_node is None:
+            raise RuntimeError("No Laplace solution seems to have been loaded.")
```
[the remark this answers is not in the document]
Comment (ebrahimebrahim) [reply Modules/Scripted/VPAWVisualize/VPAWVisualize.py:916]: Good point. Though I suppose one could be loaded but we just failed to detect it. I'll go with "Could not find a loaded laplace solution image"

File: Modules/Scripted/VPAWVisualize/VPAWVisualize.py [hunk 3]
```diff
@@ -1,11 +1,15 @@
 import logging
 import numpy as np
 import os
+from pathlib import Path
```
[the remark this answers is not in the document]
Comment (ebrahimebrahim) [reply Modules/Scripted/VPAWVisualize/VPAWVisualize.py:4]: I am against unqualified import like `from pathlib import *` for similar reasons to what you are describing, but I think the explicit `from pathlib import Path` is okay. I agree with the spirit of always goes for module level imports like `import blah`, but sometimes I think I still prefer the tradeoff towards having the shorter object name.

**Comment by ebrahimebrahim** [comment]
@Leengit Thanks for the review!

**Pull request merged**